### ISSUE TITLE
Reading in multiple root files into pandas/dask DataFrame

### ISSUE THREAD

**Issue by Popinaodude** [scikit-hep/uproot4#129]
Hi, I need to read in multiple root files and convert them into a single pandas or dask dataframe, and I am confused about the implementation  of `uproot.concatenate()` or `uproot.iterate()` and the difference between uproot 3 and 4 in this case. 

if I have inputs like:
```
root_files = 'data/*.root'
branches = ['branch1','branch2','branch3']
```

I've tried `df_generator = uproot4.iterate(root_files, treepath='my_TTree', branches=branches, outputtype=pd.Dataframe)`
and I get the following error:
`TypeError: 'files' must include a TTree/TBranch object path (separated by a colon ':') to each glob pattern (if multiple are given)`

uproot4 doesn't seem to have a pandas.iterate method so if I do `df_generator = uproot.pandas.iterate(root_files, treepath='my_TTree', branches=branches)`

I can then generate a dataframe by doing something like:
```
df_list = [data for data in df_generator]
df = pd.concat(df_list)
```
However this is incredibly slow due to the time it takes to extract each dataframe separately into the list and takes up double the memory it needs to. What would you say is the most efficient way of concatenating multiple root files into a single dataframe? Especially if the end goal is to eventually parallelise with dask or another similar library for large files

**Comment by jpivarski** [comment]
The colon syntax for a tree path and the way to select Pandas is:

```python
df_generator = uproot4.iterate("data/*.root:my_TTree", branches=branches, library="pd")
```

Absorbing the `treepath` into the file names is to allow different sets of files to have different tree paths: `["old-data/*.root:old_tree_path", "new-data/*.root:new_tree_path"]` and if you have a colon in your file names, there's `{"data/*.root": "my_TTree"}` syntax. [All of this is documented.](https://uproot4.readthedocs.io/en/latest/uproot4.behaviors.TBranch.iterate.html)

Uproot 3 has a `outputtype` to describe the type of output container—dict, list, tuple, or Pandas DataFrame—but there's a more important distinction: which array library are you using to describe the arrays? So `library="np"` returns strict NumPy, even if inefficient for jagged arrays, `library="ak"` (default) returns Awkward Arrays, and `library="pd"` is for Pandas: Series or DataFrames depending on `array` vs `arrays` (in your case, it will be a DataFrame). There's also a `how` parameter for how you want that grouped: dict, list, tuple, or a merging strategy for Pandas if you have branches of different jaggedness. (`how` is passed straight into Pandas's merging `how`: it's a string like `"inner"` or `"outer"` for inner and outer joins.)

Given that your eventual goal is to concatenate everything, they'll have to fit into memory, but also, there's an [uproot4.concatenate](https://uproot4.readthedocs.io/en/latest/uproot4.behaviors.TBranch.concatenate.html) function for that. Like iterate, it selects between Pandas and other libraries with a `library` parameter. I think it uses Pandas's own `concat` function to do the concatenation. If you've already found that that's too slow/memory hogging, maybe you want to pass `library="np"` or `library="ak"` to `uproot4.concatenate` and then convert the final result to Pandas with [ak.to_pandas](https://awkward-array.readthedocs.io/en/latest/_auto/ak.to_pandas.html) or (in NumPy's case) the Pandas DataFrame constructor.

If your data are flat (non-jagged, all of your selected branches are one-value-per-event, not `std::vector` or anything), then NumPy is the fastest thing around. Awkward Array and Pandas _use_ NumPy when they can and anything they need to implement separately can't match NumPy's decades of tuning. If your data are jagged, Awkward Array is your best bet, because it's the only one of the three that was designed for that. Pandas, generally speaking, is slow and memory-hogging. (See [its author's rant](https://wesmckinney.com/blog/apache-arrow-pandas-internals/).)

**Comment by Popinaodude** [comment]
Thank you for the reply! I was actually really struggling to find the uproot4 docs as they're not in the git readme I just naively assumed they hadn't been done yet. 
I'll be taking this all into consideration. I think for now I'll stick with pandas as I need to be very flexible in which and how many branches I am reading in but if I reach a roadblock I now know I should probably be switching back to numpy. Thanks again

**Comment by jpivarski** [comment]
(I can't believe I forgot to put the full documentation into the GitHub README! Thanks!)

**Comment by jpivarski** [comment]
Fixed by #133.
